akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala

```diff
@@ -46,7 +46,41 @@ class SourceSpec extends StreamSpec with DefaultTimeout {
       p.subscribe(c2)
       c2.expectSubscriptionAndError()
     }
+  }
+
+  "Future Source" must {
+    import akka.stream.impl.fusing.GraphStages.FutureFlattenSource
+
+    "flatten elements" in Utils.assertAllStagesStopped {
+      implicit def ec = materializer.executionContext
+
+      val subSource: Source[Int, String] =
+        Source(List(1, 2, 3)).mapMaterializedValue(_ ⇒ "foo")
+
+      val futureSource = new FutureFlattenSource(Future(subSource))
+      val source: Source[Int, Future[String]] = Source.fromGraph(futureSource)
+
+      val materialized = Promise[String]()
+      val watched: Source[Int, NotUsed] = source.watchTermination() { (m, d) ⇒
+        materialized.completeWith(d.flatMap(_ ⇒ m))
+        NotUsed
+      }
+
+      val p = watched.runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[Int]()
+      p.subscribe(c)
+
+      val sub = c.expectSubscription()
+      sub.request(5)
+
+      c.expectNext(1)
+      c.expectNext(2)
+      c.expectNext(3)
 
+      c.expectComplete()
+
+      Await.result(materialized.future, 3.seconds) should ===("foo")
+    }
```

Comment:
I think the tests for this should be in a separate `FutureFlattenSourceSpec`. It also needs considerably more test coverage. Example use cases: what happens with the stream and materialized value if

* downstream cancels before the future is completed
* the future is completed with a failure
* downstream is applying backpressure when the future completes
* the future is completed with a graph that fails to materialize (throws exception)